machine_translation/sampling.py

```diff
@@ -0,0 +1,301 @@
+import logging
+import numpy
+import operator
+import os
+import re
+import signal
+import time
+
+from blocks.extensions import SimpleExtension
+from blocks.search import BeamSearch
+
+from subprocess import Popen, PIPE
+
+logger = logging.getLogger(__name__)
+
+
+class SamplingBase(object):
+    """Utility class for BleuValidator and Sampler."""
+
+    def _get_attr_rec(self, obj, attr):
+        return self._get_attr_rec(getattr(obj, attr), attr) \
+            if hasattr(obj, attr) else obj
+
+    def _get_true_length(self, seq, vocab):
+        try:
+            return seq.tolist().index(vocab['</S>']) + 1
+        except ValueError:
+            return len(seq)
+
+    def _oov_to_unk(self, seq):
+        return [x if x < self.config['src_vocab_size'] else self.unk_idx
+                for x in seq]
+
+    def _parse_input(self, line):
```

Comment:
fixed by  cd9eacc. @fhirschmann we actually do need padding, but not specifically zero padding. In general we pad with end of sequence idx for convenience. `_parse_input` was not being used hence i removed it, nice catch again :) thanks.